File: src/DistributionallyRobustMain.java
import java.util.Random;

public class DistributionallyRobustMain {
    public static void main(String[] args) throws Exception {
        // 指定输入文件和输出文件名
        String instanceFile = "./Instances/2DU60-05-1.dat";  // 使用与ChanceConstrainedMain类似的实例
        String outputFileName = "2DU60-05-1_drcc";

        // 设置分布鲁棒优化参数
        double gamma = 0.1;  // 风险参数（违反概率）
        int numScenarios = 100;  // 场景数量
        double r = 0.1;
        long seed = 12345678;  // 随机种子

        // 需求期望值和相对标准差
        double E = 50.0; // 期望值
        double[] RSDValues = {0.125, 0.25, 0.5}; // 相对标准差数组

        // 选择特定的RSD值
        double RSD = RSDValues[0]; // 可以通过索引选择不同的RSD值

        // 分布鲁棒优化其他参数
        boolean useD1 = true;  // 是否使用D_1模糊集（false则使用D_2模糊集）
        double delta1 = 2;  // D_2模糊集参数
        double delta2 = 4;   // D_2模糊集参数（需保证delta2 > max{delta1, 1}）
        boolean useJointChance = false;  // 是否使用联合机会约束（Bonferroni近似）

        // 加载实例
        Instance instance = new Instance(instanceFile);

        // 生成随机场景 - 使用均匀分布
        double[][] scenarios = generateScenarios(instance.getN(), numScenarios, E, RSD, seed);


        // 默认使用近似方法（MISOCP），可以通过参数修改
        boolean useExactMethod = false; // 设置为true使用精确方法（割约束）
        
        DistributionallyRobustAlgo algo = new DistributionallyRobustAlgo(
                instance, scenarios, gamma, seed, useD1, delta1, delta2, useJointChance, r, useExactMethod);

        // 运行算法并生成结果
        algo.run(outputFileName);

        System.out.println("基于分布鲁棒机会约束的配送区域划分问题求解完成。");

        // 可视化结果
        String outputImagePath = "./output/" + outputFileName + "_visualization.png";
        DistrictVisualizer visualizer = new DistrictVisualizer(instance, algo.getZones(), algo.getCenters());
        visualizer.saveVisualization(outputImagePath);

        System.out.println("可视化图像已保存至: " + outputImagePath);
    }

    // 生成随机场景 - 使用均匀分布
    private static double[][] generateScenarios(int n, int numScenarios, double E, double RSD, long seed) {
        double[][] scenarios = new double[numScenarios][n];
        Random rand = new Random(seed); // 固定种子以保证可重复性

        // 计算均匀分布的左右端点
        double lowerBound = E * (1 - Math.sqrt(3) * RSD);
        double upperBound = E * (1 + Math.sqrt(3) * RSD);

        // 生成场景
        for (int s = 0; s < numScenarios; s++) {
            for (int i = 0; i < n; i++) {
                // 均匀分布生成需求
                double demand = lowerBound + rand.nextDouble() * (upperBound - lowerBound);

                // 确保需求为正值
                scenarios[s][i] = Math.max(1, demand);
            }
        }

        return scenarios;
    }

    // 比较不同模型的性能
//    private static void runComparativeAnalysis(Instance instance, double[][] scenarios,
//                                               double gamma, long seed, String baseFileName) throws Exception {
//        System.out.println("\n开始进行比较分析...");
//
//        // 1. 运行确定性模型
//        Algo deterministicAlgo = new Algo(instance);
//        deterministicAlgo.run(baseFileName + "_det");
//
//        // 2. 运行场景近似模型
//        ChanceConstrainedAlgo ccaAlgo = new ChanceConstrainedAlgo(instance, scenarios, gamma, seed,0.1);
//        ccaAlgo.run(baseFileName + "_cc");
//
//        // 3. 运行D_2模糊集分布鲁棒模型
//        DistributionallyRobustAlgo drAlgo = new DistributionallyRobustAlgo(
//                instance, scenarios, gamma, seed, false, 0.05, 0.5, false);
//        drAlgo.run(baseFileName + "_dr_d2");
//
//        // 输出各模型的目标函数值和求解时间等信息
//        System.out.println("比较分析完成。详细结果请参见输出文件。");
//    }
}

File: MatlabCode/DistributionallyRobustMain.m

% DistributionallyRobustMain.m
function DistributionallyRobustMain()
    clc;clear;
    % Start timer for overall execution
    totalStartTime = tic;
    timerActive = false;
    timeoutTimer = [];
    
    try
        % Set parameters
        instanceFile = './Instances/2DU60-05-1.dat';  % Input file path
        outputFileName = 'test_drcc';           % Output file name
        outputDir = './output/';
        
        % Set distributionally robust optimization parameters
        gamma = 0.2;      % Risk parameter
        numScenarios = 1000; % Number of scenarios
        seed = 12345678;   % Random seed
        
        % Demand parameters
        E = 50.0;          % Expected value
        RSDValues = [0.125, 0.25, 0.5]; % Relative standard deviation array
        
        % Select specific RSD value
        RSD = RSDValues(1);
        r = 0.2;
        % Other DRCC parameters
        useD1 = true;      % Whether to use D1 fuzzy set
        delta1 = 1;        % D2 fuzzy set parameter
        delta2 = 2;        % D2 fuzzy set parameter
        useJointChance = false; % Whether to use joint chance constraint
        
        % Algorithm timeout (in seconds)
        timeout = 3600; % 1 hour timeout
        
        % Display configuration
        % displayConfig(instanceFile, outputFileName, gamma, numScenarios, E, RSD, useD1, delta1, delta2, useJointChance);
        
        % Validate parameters
        % validateParams(gamma, numScenarios, E, RSD, delta1, delta2);
        
        % Ensure output directory exists
        if ~exist(outputDir, 'dir')
            [success, msg] = mkdir(outputDir);
            if ~success
                error('Failed to create directory %s: %s', outputDir, msg);
            end
            fprintf('Created output directory: %s\n', outputDir);
        end
        
        % Check if instance file exists
        if ~exist(instanceFile, 'file')
            error('Instance file not found: %s', instanceFile);
        end
        
        % Load instance
        fprintf('Loading instance from %s...\n', instanceFile);
        instance = Instance(instanceFile);
        
        if isempty(instance) || instance.getN() == 0
            error('Failed to load instance or instance has no areas');
        end
        
        % fprintf('Instance contains %d areas\n', instance.getN());
        
        % Generate random scenarios using vectorized operations
        % fprintf('Generating %d scenarios...\n', numScenarios);
        scenarios = generateScenarios(instance.getN(), numScenarios, E, RSD, seed);
        
        % Create and run the distributionally robust algorithm
        % fprintf('Initializing algorithm...\n');
        algo = DistributionallyRobustAlgo(instance, scenarios, gamma, seed, useD1, delta1, delta2, useJointChance,r);
        
        if isempty(algo)
            error('Failed to create algorithm object');
        end
        
        % Try to set algorithm timeout
        try
            if isfield(algo, 'timeLimit') || isprop(algo, 'timeLimit')
                algo.timeLimit = timeout;
                % fprintf('Algorithm timeout set to %d seconds\n', timeout);
            end
        catch
            fprintf('Warning: Could not set algorithm timeout\n');
        end
        
        % Run algorithm with timeout monitoring
        % fprintf('\nRunning algorithm...\n');
        algoRunStartTime = tic;
        
        % Try to create a timer for timeout monitoring
        try
            if exist('timer', 'file') && exist('timerfind', 'file')
                % Delete any existing timers with the same name
                existingTimers = timerfind('Name', 'TimeoutTimer');
                if ~isempty(existingTimers)
                    delete(existingTimers);
                end
                
                % Create a new timer
                timeoutTimer = timer('TimerFcn', @(~,~) timeoutCheck(algoRunStartTime, timeout), ...
                       'ExecutionMode', 'fixedRate', ...
                       'Period', 10, ... % Check every 10 seconds
                       'Name', 'TimeoutTimer');
                start(timeoutTimer);
                timerActive = true;
            end
        catch timerError
            fprintf('Warning: Could not create timeout timer: %s\n', timerError.message);
        end
        
        % Run algorithm
        algo.run(outputFileName);
        
        % Stop the timer if it's active
        cleanupTimer();
        
        fprintf('Distributionally robust chance-constrained district design problem solved.\n');
        
        % Get results from the algorithm
        zones = algo.getZones();
        centers = algo.getCenters();
        % Perform out-of-sample performance evaluation
        fprintf('\nPerforming out-of-sample performance evaluation...\n');
        evaluateOutOfSamplePerformance(instance, zones, centers, E, RSD,r);

% Store performance results in output if needed
% You could write these results to a file if desired
        % Visualize results
        
        % Display solution statistics
        % Total runtime
        % totalRunTime = toc(totalStartTime);
        % fprintf('\nTotal run time: %.2f seconds (%.2f minutes)\n', ...
        %     totalRunTime, totalRunTime/60);
        
    catch e
        % Stop the timer if it's active
        cleanupTimer();
        
        fprintf('\n\n');
        fprintf('Error occurred: %s\n', e.message);
        fprintf('Stack trace:\n');
        for i = 1:length(e.stack)
            fprintf('  File: %s, Line: %d, Function: %s\n', ...
                e.stack(i).file, e.stack(i).line, e.stack(i).name);
        end
    end
    
    % Nested function to clean up timer
    function cleanupTimer()
        if timerActive && ~isempty(timeoutTimer) && isvalid(timeoutTimer)
            try
                stop(timeoutTimer);
                delete(timeoutTimer);
            catch
                % Ignore errors in timer cleanup
            end
            timerActive = false;
        end
    end
end

% Function to display configuration
function displayConfig(instanceFile, outputFileName, gamma, numScenarios, E, RSD, useD1, delta1, delta2, useJointChance)
    fprintf('----------------------------------------\n');
    fprintf('Distributionally Robust Optimization Configuration:\n');
    fprintf('----------------------------------------\n');
    fprintf('Instance file: %s\n', instanceFile);
    fprintf('Output file: %s\n', outputFileName);
    fprintf('Risk parameter (gamma): %.4f\n', gamma);
    fprintf('Number of scenarios: %d\n', numScenarios);
    fprintf('Expected demand value (E): %.2f\n', E);
    fprintf('Relative standard deviation (RSD): %.4f\n', RSD);
    
    if useD1
        fprintf('Fuzzy set type: D1\n');
    else
        fprintf('Fuzzy set type: D2\n');
        fprintf('Delta1 parameter: %.2f\n', delta1);
        fprintf('Delta2 parameter: %.2f\n', delta2);
    end
    
    if useJointChance
        fprintf('Constraint type: Joint chance constraint\n');
    else
        fprintf('Constraint type: Individual chance constraint\n');
    end
    fprintf('----------------------------------------\n\n');
end

% Function to validate parameters
function validateParams(gamma, numScenarios, E, RSD, delta1, delta2)
    % Check gamma (risk parameter)
    if ~isnumeric(gamma) || ~isscalar(gamma) || gamma <= 0 || gamma >= 1
        error('Parameter "gamma" must be a scalar between 0 and 1');
    end
    
    % Check numScenarios
    if ~isnumeric(numScenarios) || ~isscalar(numScenarios) || numScenarios <= 0 || mod(numScenarios, 1) ~= 0
        error('Parameter "numScenarios" must be a positive integer');
    end
    
    % Check E (expected value)
    if ~isnumeric(E) || ~isscalar(E) || E <= 0
        error('Parameter "E" must be a positive scalar');
    end
    
    % Check RSD (relative standard deviation)
    if ~isnumeric(RSD) || ~isscalar(RSD) || RSD <= 0
        error('Parameter "RSD" must be a positive scalar');
    end
    
    % Check delta1 and delta2 (D2 fuzzy set parameters)
    if ~isnumeric(delta1) || ~isscalar(delta1) || delta1 <= 0
        error('Parameter "delta1" must be a positive scalar');
    end
    
    if ~isnumeric(delta2) || ~isscalar(delta2) || delta2 <= 0
        error('Parameter "delta2" must be a positive scalar');
    end
    
    if delta2 <= delta1
        error('Parameter "delta2" must be greater than "delta1"');
    end
    
    % Check if lower bound would be negative
    lowerBound = E * (1 - sqrt(3) * RSD);
    if lowerBound <= 0
        warning(['With E=%.2f and RSD=%.4f, the calculated lower bound (%.2f) is non-positive. ' ...
                'This will be adjusted to 1 for demand generation.'], E, RSD, lowerBound);
    end
end

% Function to generate scenarios - vectorized version
function scenarios = generateScenarios(n, numScenarios, E, RSD, seed)
    % Set random seed for reproducibility
    rng(seed);
    
    % Calculate uniform distribution endpoints
    lowerBound = E * (1 - sqrt(3) * RSD);
    upperBound = E * (1 + sqrt(3) * RSD);
    
    % Handle potential negative lower bound
    if lowerBound <= 0
        warning('Lower bound for demand is non-positive. Setting to 1.');
        lowerBound = 1;
    end
    
    % Initialize scenarios matrix
    scenarios = zeros(numScenarios, n);
    
    % For very large instances, generate in chunks to save memory
    maxChunkSize = 1000; % Adjust based on available memory
    
    % If matrix is very large, generate in chunks
    if numScenarios * n > 10^7 % Threshold for chunk processing
        fprintf('Large instance detected. Generating scenarios in chunks...\n');
        
        for chunkStart = 1:maxChunkSize:numScenarios
            % Determine chunk end
            chunkEnd = min(chunkStart + maxChunkSize - 1, numScenarios);
            chunkSize = chunkEnd - chunkStart + 1;
            
            % Generate random values for this chunk
            randomValues = rand(chunkSize, n);
            
            % Scale to the desired range
            scenarios(chunkStart:chunkEnd, :) = lowerBound + randomValues * (upperBound - lowerBound);
            
            % Report progress
            fprintf('Generated scenarios %d to %d of %d\n', chunkStart, chunkEnd, numScenarios);
        end
    else
        % For smaller instances, generate all at once
        randomValues = rand(numScenarios, n);
        scenarios = lowerBound + randomValues * (upperBound - lowerBound);
    end
    
    % Ensure all demands are positive (at least 1)
    scenarios = max(1, scenarios);
end

% Function to check if algorithm execution has exceeded timeout
function timeoutCheck(startTime, timeout)
    elapsedTime = toc(startTime);
    if elapsedTime > timeout
        % This will be caught by the try-catch block in the main function
        error('Algorithm execution exceeded timeout of %d seconds', timeout);
    end
end


% Function to evaluate out-of-sample performance
function performanceResults = evaluateOutOfSamplePerformance(instance, zones, ~, E, RSD, r)
    % If r is not provided, use default value 0.1
    if nargin < 6
        r = 0.1;
    end

    % fprintf('\n----------------------------------------\n');
    % fprintf('Evaluating Out-of-Sample Performance:\n');
    % fprintf('----------------------------------------\n');
    % 
    % Parameters for out-of-sample testing
    numTestScenarios = 1000;
    testSeed = 54321; % Different seed for test data
    
    % Generate test scenarios
    % fprintf('Generating %d test scenarios...\n', numTestScenarios);
    testScenarios = generateScenarios(instance.getN(), numTestScenarios, E, RSD, testSeed);
    
    % Initialize counters for each district and overall performance
    districtViolationCounts = zeros(1, length(zones));
    scenariosSatisfied = 0;
    
    % Evaluate each test scenario
    % fprintf('Evaluating scenarios...\n');
    
    % Pre-compute zone assignment matrix (each column represents a zone)
    n = instance.getN();
    assignmentMatrix = zeros(n, length(zones));
    for i = 1:length(zones)
        if ~isempty(zones{i})
            assignmentMatrix(zones{i}, i) = 1;
        end
    end
    
    % For each scenario
    for s = 1:numTestScenarios
        % Calculate total demand for this specific scenario
        scenarioTotalDemand = sum(testScenarios(s, :));
        
        % Calculate demand upper bound for this specific scenario
        scenarioDemandUpperBound = (1 + r) * (scenarioTotalDemand / instance.k);
        
        % Calculate all zone demands in one matrix operation
        scenarioDemandVector = testScenarios(s, :);
        zoneDemands = scenarioDemandVector * assignmentMatrix;
        
        % Check violations for all zones at once
        violationMask = zoneDemands > scenarioDemandUpperBound;
        districtViolationCounts = districtViolationCounts + violationMask;
        
        % Scenario is satisfied if no violations
        scenarioSatisfied = ~any(violationMask);
        if scenarioSatisfied
            scenariosSatisfied = scenariosSatisfied + 1;
        end
    end
    
    % Calculate satisfaction rates
    districtSatisfactionRates = 1 - (districtViolationCounts / numTestScenarios);
    overallSatisfactionRate = scenariosSatisfied / numTestScenarios;
    
    % Output results
    % fprintf('\nOut-of-Sample Performance Results:\n');
    fprintf('  Overall satisfaction rate: %.4f\n', overallSatisfactionRate);
    performanceResults = overallSatisfactionRate;
    % fprintf('\nIndividual district satisfaction rates:\n');
    
    % for i = 1:length(zones)
    %     fprintf('  District %d: %.4f\n', i, districtSatisfactionRates(i));
    % end
    
    % fprintf('\nMin district satisfaction rate: %.4f\n', min(districtSatisfactionRates));
    % fprintf('----------------------------------------\n');
    % 
    % Return results as a structure
    % performanceResults = struct(...
    %     'overallSatisfactionRate', overallSatisfactionRate, ...
    %     'districtSatisfactionRates', districtSatisfactionRates, ...
    %     'numTestScenarios', numTestScenarios);
end